[ADR] protocol versions and chain descriptions

ebuchman opened the pull request.

Liamsi commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+how much that matters compared to being able to upgrade the protocol.
+
+XXX: if needed, can we change the meaning of the first byte of the first message to encode a handshake version?
+this is the first byte of a 32-byte ed25519 pubkey.
```
Not sure, if we should include a pre-handshake protocol negotiation step? Similar to what TLS is doing? Let's say we change the secret connection implementation. For some time we'd allow nodes to "speak" several protocols / suites to establish an auth. enc. handshake? Of course that leaks some information but I do think being able to upgrade this (and maybe allowing peers to catch up to the latest upgrade for some time), is important.

cc @ValarDragon

ValarDragon replied:
Thanks for the cc.

Definitely agree, we should have a protocol negotiation step similar to TLS. We definitely want to be able to upgrade the encryption suite. We don't really have to worry about the additional round being added here, since its not important that its super fast to connect to a peer. 

There isn't information leakage if you're willing to connect to anyone in the p2p layer. If you're not, we should have a separate whitelisting system for the first message, that we can build post-launch.

jaekwon replied:
I think we should negotiate after a secret connection was established, ideally, rather than the other way around, just so we leak less information.

The whitelisting I think is orthogonal to making things perfect-forward-secret as much as possible.  Even if whitelisted, I'd want to use secret_connection to negotiate the version.

I don't think it matters too much in the grand scheme of things, but I see no reason to use anything other than secret_connectin at the base layer and build on top.

ebuchman replied:
I believe the concern is if we need to upgrade the algorithms used in the secret connection, if we dont negotiate before hand, we won't be able to

Liamsi replied:
Thanks for making this clearer @ebuchman. Exactly.

ValarDragon replied:
Talked to @zmanian , he had a really good point. We should make the negotiation step have the  handshake + session encryption algorithm. This way its easy to switch to noise in the future.

ValarDragon commented:
markdown rendered version link: 
https://github.com/tendermint/tendermint/blob/28bd88a97274b4a59c11bd16f3d15a86bd18b023/docs/architecture/adr-versions.md

ValarDragon commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+- `("upgrade_chain_description", <new ChainDescription>)`
+- `("upgrade_chain_height", <height to upgrade>)`
+
+This provides a simple manner for applications to signal to external processes that they want to upgrade, and what height to do it at.
```
Adding `SHA2(.tar of the source code)` makes this implementation the canonical implementation though. (We'd have to specify our language / implementation in order for this to be useful in the context of other implementations)

ebuchman replied:
Yeh we want the AppVersion to be a software/implementation agnostic version that defines only the protocol

alexanderbez commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+Compatibility requirements at the peer layer have already been described.
+
+The only other change necessary is an `/unsafe_stop?height=_` endpoint to tell Tendermint to shutdown at a given height.
```
This should be a protected endpoint, no?

cwgoes replied:
That's implied by the `unsafe` prefix.

alexanderbez commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+### Process Manager
+
+A simple process managers subscribes to the `upgrade_chain` events and calls `/unsafe_stop` for the given height.
```
Just to be clear, this is to avoid hard-coded conditionals on upgrades via app versions correct? Will this just be a service that runs along side a Tendermint node?

cwgoes commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+    - this means we checkout different versions of the software at different heights instead of littering the code
+          with conditionals
+    - minimize the number of data migrations required across AppVersion (ie. most AppVersion should be able to read the same state from disk as previous AppVersion).
+- NOTE: we still need to support peers syncing from old heights across AppVersion changes
```
If we want to do this, I think we need something like https://github.com/cosmos/cosmos-sdk/issues/1602, which *would* involve having multiple code versions (although no littered conditionals).

cwgoes commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+P2PVersion is the first field in the NodeInfo. NodeInfo is also proto3 so this is easy to read out.
+
+For each P2PVersion, we keep a list of previous versions that this one is compatible with so we can connect to
```
Can we use a semantic version?

ebuchman replied:
Semantic version doesn't cut it - we want to be able to make a sequence of non-breaking changes that result in a breaking change. ie. v20 is compatible with v19, and v21 is compatible with v20, but v21 is not compatible with v19

ValarDragon replied:
Agreed for not using semver, but for a different reason.

There may be multiple types of p2p broadcasts you want to support that are entirely distinct. I think we should list the supported p2p broacst methods, and both nodes list all the things that they support. It may be the case that eventually people want to support fancier broadcast methods. I recall an issue where zarko and jeremiah were discussing it, however I can't find that issue rn.



Maybe there should be a secret connection negotiation step, and a tx broadcast method (After authentication) negotiation step.

cwgoes commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+Compatibility requirements at the peer layer have already been described.
+
+The only other change necessary is an `/unsafe_stop?height=_` endpoint to tell Tendermint to shutdown at a given height.
+This is to enable it to be stopped at the height signalled by `upgrade_chain_height` and to then be restarted using the new ChainDescription (ie. new app).
```
I wonder if there's a way we can dynamically link the ABCI application so that Tendermint itself doesn't need to be restarted for state machine changes. Might be difficult.

cwgoes commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+Note because of Tendermint's voting power threshold rules, a chain can only be extended under the "original" rules and under the new rules
+if 1/3 or more is double signing, which is expressly prohibited, and is supposed to result in their punishment on both chains. Since they can censor
+the punishment, the chain is expected to be hardforked to remove the validators. Thus, if both branches are to continue after a fork,
+they will each require a new identifier, and the old chain identifier will be retired (ie. only useful for syncing history, not for new blocks)..
```
I'm not sure this is what will actually happen in a "contentious fork", either or both "sides" may elect to have any validators who sign on the other chain be slashable on their chain.

Of course, "contentious" forks may disregard versioning rules anyways.

kansi replied:
It is not really clear how will slashing work in case the forks adopt new chain identifiers? Since the new blocks belong to different chains so two blocks at a given height `h` belonging to different chains signed by the same validator set cannot be considered as a reason to slash? I would be a great help if a little more detail could be added to this.

ebuchman replied:
great point. chains will have to remember their "original" chain id so they can punish. I'll add a note

cwgoes commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+- Behaviour and message types will change abruptly in the course of the life of a chain
+- Need to minimize complexity of the code for supporting different AppVersions at different heights
+- Ideally, each version of the software supports only a *single* AppVersion at one time
+    - this means we checkout different versions of the software at different heights instead of littering the code
```
It would be *possible* to integrate Git and the Golang compiler into the SDK for actually "automatic" upgrades in this vein.

Not sure if it's worth the complexity (and not requiring manual upgrades introduces some risk).

ValarDragon replied:
That would lock us into golang being the canonical language. I don't think we want that. (We should want other language implementations)

I also think having every update to golang or git on chain is a bad idea.

cwgoes replied:
I don't think that's true - it would make it *harder* to write implementations in other languages, but one could write the same kind of automation with another compiler (perhaps including in the upgrade two Git hashes for the different languages).

You would need a *lot* of protocol upgrades for the code size to be significant - at the moment, the SDK has ~8 MB of Go source (and it would be compressed in Git).

I agree that this is complex though.

cwgoes replied:
The source code can be compressed, which will cut that by a factor of 2-3 perhaps. 

Generally, though, I agree - this is definitely a post-launch consideration, and even then may not be worth the complexity.

needkane commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+The Software Version is covered by SemVer and described elsewhere.
+It is not relevant to the protocol description, suffice to say that if any protocol version
+changes, the software version changes. Software version is included in NodeInfo for convenience/diagnostics.
```
Now update version must change code： https://github.com/tendermint/tendermint/commit/2fa99628351bd6d4ba14226bb7336135d655ff40.


There is a problem, if  already run old version(major  version different) nodes，and many user use them everyday. I can't upgrade one node which use new version code. 

In my opinion , new version should compatible old version. At least it should not be due to the issue of version number, such as Ethereum 、 Bitcoin

ebuchman replied:
Once we introduce protocol versions, the software version won't be relevant for peer connectivity - just the protocol versions.

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
+- To effectively change it, dump state and start a new blockchain with the new BlockVersion.
```
We need to be more precise what happens when Protocol version change. Does this mean that we force all nodes to update software in a coordinated way?

ebuchman replied:
It means we're starting a new blockchain from the latest state of the old one with a new version (like what we've been doing on testnets).

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+- Will change gradually as reactors evolve to improve performance and support new features
+	- eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
+- It's easy to determine the version of a peer from its first serialized message/s
+- New versions must be compatible with at least one old version to allow gradual upgrades
```
I don't understand this. Does this mean that the changes between two versions should always be done in a backward compatible way.

ValarDragon replied:
I'd prefer we not require compatibility with old versions. That seems kind of unnatural to me, plus its unenforceable. I think listing all the previous versions a node supports would be more ideal.

ebuchman replied:
> Does this mean that the changes between two versions should always be done in a backward compatible way.

For the p2p layer, yes. It means you can upgrade at least once and still connect successfully to the rest of the network.

Not sure why we wouldn't want that requirement at the peer layer. 

> I think listing all the previous versions a node supports would be more ideal.

This is what we're proposing, and requiring the list be non-empty

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+#### AppVersion
+
+- The ABCI state machine (txs, begin/endblock behaviour, commit hashing)
```
Why we care about this at Tendermint level? Isn't Tendermint app agnostic? Even for state syncing, we need to have a mechanisms that transfer bytes around but we don't care what is inside. It is app business.

ValarDragon replied:
I think we want to reject peers with a different app version

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+They must have distinct identifiers so that peers know which one they are joining and so
+validators and users can prevent replay attacks.
+
+Call this the `NetworkName`. It represents both the application being run and the community or intention
```
What is different between chain id and network name?

ebuchman replied:
What we currently call "ChainID" in the software is what I'm here calling "NetworkName"

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+Peers only connect to other peers with the same NetworkName.
+
+### Forks
```
Why we discuss forks here? I thought fork means a new chain, and we do versioning only in the context of a single chain.

ebuchman replied:
This proposal considers making forks part of versioning - in some sense, upgrading the version may be a fork. But it's actually quite subtle because double-signing detection is based on ChainID so we need to consider that maybe we don't want the ChainID to change just because the version changed. So this whole ChainDescription proposal probably needs more thought. Perhaps I should split this ADR into two parts - one to introduce the basic versions, and one for the ChainDescription and Fork stuff.

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+### Versions
+
+Each of BlockVersion, AppVersion, P2PVersion is a monotonically increasing int64.
```
It seems that we assume that versions will never be backward compatible, i.e., that every new version will be breaking. Why we don't we use semantic versioning for versions? At least for P2PVersion we could probably have protocol change that is backward compatible (for example performance improvement that adds new message types, but that can still work with the old versions).

ebuchman replied:
Semantic version doesn't cut it - we want to be able to make a sequence of non-breaking changes that result in a breaking change. ie. v20 is compatible with v19, and v21 is compatible with v20, but v21 is not compatible with v19.

This proposal is assuming for simplicity that the Block and App protocol versions are never backwards compatible

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+BlockVersion is the first field in the block Header. Since we have settled on a proto3 header,
+the ability to read the BlockVersion out of the serialized header is unanimous.
+
+BlockVersion is included in the NodeInfo and we only connect to peers with the same or higher BlockVersion.
```
It would be nice to do a mental exercise of various scenarios where we have some version change at subset of nodes and then analyse what happens in the network (at the node level).

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+Note we need the peer/reactor protocols to take the versions of peers into account when sending messages:
+    - don't send messages they don't understand
+    - don't send messages they don't expect
```
This is probably also the case for Blockchain versions. If you change protocol by adding new message type, then you shouldn't sent it to a peer who does not understand it (it is at the smaller version).

ebuchman replied:
adding a new message type is a change to the P2PVersion

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+AppVersion is also included in the block Header and the NodeInfo.
+
+We only maintain connections to peers with the correct AppVersion for the height they're at.
```
I don't understand this. Can a peer update app version but stay with the same Tendermint version? Why we shouldn't connect to that peer?

ebuchman replied:
Correct. Because an updated AppVersions means they will execute transactions differently so we will not be able to agree on the results of a block.

milosevic replied:
Very good point, thanks.

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+That is, we only connect to current peers with the same AppVersion, and to old peers that have the right AppVersion for where they
+are in the syncing process.
+
+Note this requires all consensus nodes to upgrade their versions at exactly the required height.
```
This is not clear. What do you mean by consensus node?

ebuchman replied:
Validators

milosevic commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+#### ChainID
+
+Define `ChainID = TMHASH(ChainDescriptor)`. It's the unique ID of a blockchain.
```
What is the usage of chainID?

ebuchman replied:
It's the ultimate and final identifier for the chain. It's included in the SignBytes for votes and proposals so they are tied to only this chain.

ValarDragon replied:
Should the ChainID be `TMHASH(ChainDescriptor || genesis state)`? Then this provides some degree of fork seperation. (Most forks would have a different genesis state, and therefore different chain id)

ebuchman replied:
ChainDescriptor already includes the state

kansi commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+Note because of Tendermint's voting power threshold rules, a chain can only be extended under the "original" rules and under the new rules
+if 1/3 or more is double signing, which is expressly prohibited, and is supposed to result in their punishment on both chains. Since they can censor
+the punishment, the chain is expected to be hardforked to remove the validators. Thus, if both branches are to continue after a fork,
+they will each require a new identifier, and the old chain identifier will be retired (ie. only useful for syncing history, not for new blocks)..
```
It seems that there would be at least 2 chain ids i.e. (`old_chain_id`, `new_chain_id1`). In order to achieve this migration, the ABCI proxy app should be made aware about the context of a block i.e. to which chain id the given block belongs so that the block can be stored appropriately. It seems that `begin_block` api call would need the chain id as an argument to facilitate the same? Furthermore, `check_tx` might also need to know the same so that the transactions can be validated against the correct chain?

ebuchman replied:
ABCI receives the header in BeginBlock which contains the ChainID

ValarDragon commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+- It should be the least frequent/likely to change.
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
```
Also I don't think its crazy to assume we want to make our data structures more space efficient post-launch. (i.e. remove unnecessary data) A year seems like too much to me as well.

ValarDragon commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
+- To effectively change it, dump state and start a new blockchain with the new BlockVersion.
```
Is this the ideal method for change of block version? The chain-id would change as a consequence of this as well (Since its determined from hash of the genesis file afaik). Is an if conditional that everyone agrees too that bad here?

ValarDragon commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+- It's easy to determine the version of a block from its serialized form
+
+#### P2PVersion
+
```
I don't think a semver number is suitable for this. There may be multiple types of p2p broadcasts you want to support. I think these should be `p2p methods` or smth, and both nodes list all the things that they support. It may be the case that eventually people want to support fancier broadcast methods. I recall an issue where zarko and jeremiah were discussing it, however I can't find that issue rn.

ValarDragon replied:
Maybe there should be a secret connection negotiation step, and a tx broadcast method (After authentication) negotiation step.

ValarDragon commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+We need to support many independent networks running the same version of the software,
+even possibly starting from the same initial state.
+They must have distinct identifiers so that peers know which one they are joining and so
```
How are we enforcing / recommending distinctness? I recall a proposal to use the hash of the genesis for the chain-id floating around at one point in time.

ebuchman commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+
+## Requirements
+
+We need to version blockchains across protocols, networks, forks, etc.
```
Requirements are broken out for each of Block/P2P/App immediately beloew

ebuchman commented on docs/architecture/adr-versions.md:
```diff
@@ -0,0 +1,219 @@
+We need to version blockchains across protocols, networks, forks, etc.
+We need to do it in a way that is scalable and maintainable - we can't just litter
+the code with conditionals.
+We need chain identifiers and descriptions so we can talk about a multitude of chains in a meaningful way.
```
Will clarify. Thinking is to have something that clearly identifies the differences between different chains (ie. in terms of versions, state root, valset, etc.)

ebuchman commented:
Thanks for all the review folks. I pushed a new commit to address most of the feedback. I think probably this ADR should be split in two - one for incorporating Block/App/P2P Versions into the software, and one for dealing with forks/chain-descriptions. I'll do that before merging.

ebuchman commented:
I split the ADR in two, and incorporated recent discussion about signalling for the next AppVersion using proposer blocks, but leaves open the question of how to get that information from App to Tendermint ...

ebuchman commented:
I'm going to merge this and other ADRs into develop and add some more structure to the architecture dir

ebuchman commented:
Decided to split this into two PRs, one for each ADR. 

- See Protocol Versions https://github.com/tendermint/tendermint/pull/2144, which has reduced its scope and should be good to merge (though we still need a decision on p2p-handshake-versioning).
- See Chain Versions https://github.com/tendermint/tendermint/pull/2145, which is still a WIP
